### ISSUE TITLE
Access defined runtime environment variables from powertools

### ISSUE THREAD

**Issue by Fahadmo** [awslabs/aws-lambda-powertools-roadmap#7]
**Is your feature request related to a problem? Please describe.**

AWS Lambda has defined runtime environment variables as shown here: https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime

In order to get the name of the variable we often need to go back to the documentation which can be frustrating.

**Describe the solution you'd like**

I would like to have something that would let me access all defined environment variables.
For example, this could be a class with a list of cached properties for each defined runtime variable.

```python
@cache 
@property
def AWS_REGION(self):
    return os.environ.get("AWS_REGION")
```

**Describe alternatives you've considered**
Currently, we just run the following for each environment variable we need.
```python
x = os.environ.get(<ENV VARIABLE NAME>)
```

**Comment by saragerion** [comment]
@Fahadmo personally speaking, I think this idea is brilliant. Thanks for sharing it with us.

AppConfig is outside the scope of this issue, but curious to hear about how do you see this functionality working together with AppConfig (another configuration source for Lambda)?
https://aws.amazon.com/blogs/mt/introducing-aws-appconfig-lambda-extension-deploying-application-configuration-serverless/
Asking this _not_ bring this AppConfig functionality into this issue, but just to understand better what is the desired DX considering that there might be more than 1 data source in the future.

**Comment by michaelbrewer** [comment]
@heitorlessa - So far i have found a couple interesting libraries as an inspiration for a simple framework for environment variables, defaults (.env or json settings) and parameters (appconfig, ssm, secrets manager etc..):

- [python decouple](https://github.com/henriquebastos/python-decouple) and [prettyconf](https://github.com/osantana/prettyconf) - what i like about these libraries is how simple and lightweight they are. Prettyconf does have an example of using SSM.
- [Starlette - config ](https://www.starlette.io/config/) - is much like python decouple.
- [dynaconf](https://www.dynaconf.com/) is super fancy and allows for nested properties and automatic type casting.

Personally i kind of likely the simpler `prettyconf` or `Starlette - config` api:

```python3
config = Configuration(loaders=[Environment(), EnvFile(filename=env_file), ParameterStore(), ...])
DEBUG = config('DEBUG', cast=bool, default=False)
DATABASE_URL = config('DATABASE_URL', cast=databases.DatabaseURL)
SECRET_KEY = config('SECRET_KEY', cast=Secret)
ALLOWED_HOSTS = config('ALLOWED_HOSTS', cast=CommaSeparatedStrings)
```

- loaders - could be environment then a json config file. Or SSM Parameter Store and then environment variables.
- cast - could be list, dict, json etc..

**Comment by ran-isenberg** [comment]
For people who use the parser and pydantic, there's already a very simple solution.
You can see the implementation here: https://isenberg-ran.medium.com/aws-lambda-environment-variables-best-practices-f760384c23ed.  I have a branch and tests ready to go if you want.
It does not require adding any new libraries or dependencies.

Pydantic also has support for secrets and loading configuration from files.

Just a thought ;)

**Comment by ran-isenberg** [comment]
So you dont want to add prettyconf or Starlette ? just implement a leaner version?

**Comment by michaelbrewer** [comment]
no. just the UX.

**Comment by michaelbrewer** [comment]
"couple interesting libraries as an inspiration for a simple framework"

**Comment by ran-isenberg** [comment]
I  respect that. 
I'm usually on the other side - I will add external dependencies if they are reliable, have good docs, the UX is great and somebody is maintaining them. If that does not exist, I will write the code myself.

**Comment by michaelbrewer** [comment]
i don't see the need for 70mb dependency i don't want for a simple environment utility which can be performs using vanilla python 3.7+. When a lambda is >5mb in size the performance drops too much for me use case.  

Hopefully v2 of powertools is closer on the roadmap to allow for even smaller lambdas.

**Comment by michaelbrewer** [comment]
Reference: https://mikhail.io/serverless/coldstarts/aws/

**Comment by ran-isenberg** [comment]
From my experience, the init time for my lambdas is in the regular time, no performance issues (and i monitor them).
Cold starts are solved by using provisioned concurrency.
And you can still add any of the other tools you mentioned above, you dont have to use pydantic. I'm sure they are not 70 megas too.

And about pydantic - I use it in so many other use cases that's it's as integral as powertools. But to each his own.

**Comment by michaelbrewer** [comment]
i just don't want to us pydantic :(

**Comment by ran-isenberg** [comment]
and that's just fine :)

**Comment by michaelbrewer** [comment]
`provisioned concurrency` is akin to switching to fargate

**Comment by michaelbrewer** [comment]
@heitorlessa - would it be possible to expand on what `Feature: Expose Lambda Env Vars to prevent additional hops to the documentation` means? Is this some kind of class that abstracts the environment variables much like the typescript version does this? Maybe once the requirements are agreed to, a "Good first issue" label could be added for contributors to look at :)

**Comment by ran-isenberg** [comment]
You can see a full implementation (with tests and full documentation) in my best practices github project.
Here's the env parser : https://github.com/ran-isenberg/aws-lambda-handler-cookbook/blob/main/service/handlers/utils/env_vars_parser.py

The schema for the env vars: https://github.com/ran-isenberg/aws-lambda-handler-cookbook/blob/main/service/handlers/schemas/env_vars.py

and a blog explaining why this is important ;)
https://www.ranthebuilder.cloud/post/aws-lambda-cookbook-environment-variables

@heitorlessa  FYI in case you want it for the extras package

**Comment by michaelbrewer** [comment]
@ran-isenberg zero dependencies? 

Coz I thought this would be part of core library, to reduce the use of static constants and cut down on the boilerplate. 

As Powertools uses environment variables for a lot of things, log level, feature toggles or overriding service names. It should just be very something and lightweight

1. No additional dependencies 
2. Some basic handling of str to bool, or str to enum, or str to int. 
3. An easier way to document these internal env names and types
4. An easier way to discover what settings are supported by the internal libraries. 

If can of coz be extended to do extra stuff in your BYO implementations. Or just to append your own set of environment variables.

**Comment by ran-isenberg** [comment]
> @ran-isenberg zero dependencies?
> 
> Coz I thought this would be part of core library, to reduce the use of static constants and cut down on the boilerplate.
> 
> As Powertools uses environment variables for a lot of things, log level, feature toggles or overriding service names. It should just be very something and lightweight
> 
> 1. No additional dependencies
> 2. Some basic handling of str to bool, or str to enum, or str to int.
> 3. An easier way to document these internal env names and types
> 4. An easier way to discover what settings are supported by the internal libraries.
> 
> If can of coz be extended to do extra stuff in your BYO implementations. Or just to append your own set of environment variables.

No, like I said, it uses Pydantic for parsing, so it can only used in the extras. For those who already use pydantic, this is a no brainer.

**Comment by michaelbrewer** [comment]
Yep. Definitely a plus for the Pydantic specific lambdas and possible an entire theme on its own. Currently it looks like the project will be on lock down until the end of July. So hopefully after that there is some directions on 

1. V2 - yah / nah
2. Modularization - yah / nah
3. Official lambda layer without boto3 (after integration testing is down)
4. What would be the way to vote / request / RFCs on new modules etc.
